awx/main/models/events.py

```diff
@@ -484,29 +487,29 @@ def _update_host_summary_from_stats(self, hostnames):
             if not self.job or not self.job.inventory:
                 logger.info('Event {} missing job or inventory, host summaries not updated'.format(self.pk))
                 return
-            qs = self.job.inventory.hosts.filter(name__in=hostnames)
             job = self.job
+
+            from awx.main.models.jobs import JobHostSummary  # circular import
+            summaries = dict()
             for host in hostnames:
                 host_stats = {}
                 for stat in ('changed', 'dark', 'failures', 'ignored', 'ok', 'processed', 'rescued', 'skipped'):
                     try:
                         host_stats[stat] = self.event_data.get(stat, {}).get(host, 0)
                     except AttributeError:  # in case event_data[stat] isn't a dict.
                         pass
-                if qs.filter(name=host).exists():
-                    host_actual = qs.get(name=host)
-                    host_summary, created = job.job_host_summaries.get_or_create(host=host_actual, host_name=host_actual.name, defaults=host_stats)
-                else:
-                    host_summary, created = job.job_host_summaries.get_or_create(host_name=host, defaults=host_stats)
-
-                if not created:
-                    update_fields = []
-                    for stat, value in host_stats.items():
-                        if getattr(host_summary, stat) != value:
-                            setattr(host_summary, stat, value)
-                            update_fields.append(stat)
-                    if update_fields:
-                        host_summary.save(update_fields=update_fields)
+                host_id = self.host_map.get(host, None)
+                summaries.setdefault(
+                    (host_id, host),
+                    JobHostSummary(created=now(), modified=now(), job_id=job.id, host_id=host_id, host_name=host)
+                )
+                host_summary = summaries[(host_id, host)]
+
+                for stat, value in host_stats.items():
+                    if getattr(host_summary, stat) != value:
+                        setattr(host_summary, stat, value)
```

Comment:
Even if I assume that `hostnames` has duplicate values, there's no way that the host summary fields will be meaningfully different. `job_id` is trivially the same, `host_id` and `host_name` are what you used to look up the record in the first place, and no one cares about updating created/modified to a few microseconds later.

It's very likely that the loop never did anything before, but I don't see any way to read this where it's necessary now.

Reply:
I was also _very_ skeptical that `hostnames` would have duplicates, but I also didn't want to remove it in case there's some arcane Ansible behavior I'm unaware of.

Reply:
Following the blame to find out when this was introduced takes many turns, and I wound up here:

52ace2700cd7f5f81d47507c78aa2a43cc9069b5

For some perspective, that was 46 days after the _first_ commit.

The commit answers my question as to _why_ this loop exists. The host summary was created in the host `.save` method when a host is created. Perhaps the author anticipated these to be updated on a rolling basis while the job was in progress. There was nothing wrong with that decision at that time. However, being in the year 2020, I know relatively confidently that host summaries are not created until the end of the job, and anything else would fairly severely violate the spirit of #6991

Reply:
@matburt you have any insights into this ^ (or thoughts on this PR in general)?